### PR TITLE
Dynamic inference context | UVM only.

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
# What does this PR do ?
Dynamic inference context `memory_buffer` only supports unified memory (i.e., GPU tensor is no longer an option).

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/engines/dynamic_engine.py [hunk 1]
```diff
@@ -58,6 +60,9 @@
     HAVE_MSGPACK = False
 
 
+process = psutil.Process(os.getpid())
```
Comment (tdene) [inline megatron/core/inference/engines/dynamic_engine.py]: Cleanup

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 2]
```diff
@@ -116,11 +115,40 @@ def get_model() -> MegatronModule:
     return model
 
 
-def get_inference_context(requests: List[Request], sampling_params: SamplingParams, 
-                          calculate_max_sequence_length_from_requests: bool =True):
+def get_inference_context_active_buffer_size_bytes(model: torch.nn.Module) -> int:
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py]: As a customer, I want to be able to pass the Megatron arg directly into the context constructor, instead of having to run `int(active_buffer_size_gb * 1024**3)` as a translation layer.

As far as I see it, that leaves two options:

1. Change the `--inference-dynamic-batching-active-buffer-size-gb` argument to `--inference-dynamic-batching-active-buffer-size-bytes`.
2. Make the DynamicContext `__init__` have two optional arguments: `active_buffer_size_gb` and `active_buffer_size_bytes`, and assert that exactly one of them is passed in.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 3]
```diff
@@ -235,53 +233,25 @@ def __init__(
...
-                warnings.warn(
-                    "Unified memory requested but not available; defaulting to GPU memory."
-                )
-                self.unified_memory_level = 0
-            else:
-                self.unified_memory_mempool = create_unified_mempool()
+        assert (
+            has_unified_memory
+        ), "CUDA unified memory must be available to use `DynamicInferenceContext`."
+        self.unified_memory_mempool = create_unified_mempool()
```
Comment (tdene) [inline megatron/core/inference/contexts/dynamic_context.py]: As discussed in more detail, it would be very useful to allow this rework to function without unified memory. If `unified_memory_level=0`, the user personally incurs the risk of causing OOMs. We don't make that the default argument, but we leave it available for apples-to-apples testing.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 4]
```diff
@@ -143,26 +152,18 @@ class DynamicInferenceContext(BaseInferenceContext):
...
         max_sequence_length (int): Max possible sequence length (prompt + output)
             that will occur.
-        buffer_size_gb (float): Total buffer size (GB), shared by main and
-            fallback contexts.
+        active_buffer_size_bytes (float): Buffer size reserved for active requests
+            that live on the GPU. The total buffer size (stored in unified memory)
+            is 2x this value, with the the other half of the buffer reserved for
+            paused requests that live on the CPU.
+        max_tokens (int): Max number of tokens to use for forward passes. This is
+            primarily limited by prefill activation memory usage.
```
Comment (tdene) [inline megatron/core/inference/contexts/dynamic_context.py]: It's a weird pattern to have this be a non-optional argument, but not have any argument in `megatron/training/arguments.py` that sets it.

I would make `max_tokens` an Optional argument with a default value. I think that's also going to be the way we want it once we implement automatic calculation of optimal `max_tokens` later on: an Optional argument that is hidden from the user, but power-users can override.

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 2 above)
Comment (lmcafee-nvidia) [reply examples/inference/gpt/gpt_dynamic_inference.py]: I got rid of `get_inference_context_active_buffer_size_bytes()` and `get_inference_context_max_tokens()`. I was starting to do too much when I first started this branch, and these methods are best left to a future PR when we actually automatic these calculations. Let me know if it looks okay now.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 3 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: I re-added this control

File: megatron/core/inference/contexts/dynamic_context.py (hunk 4 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: yeah I agree, I gave it a default value, and as I mentioned above I got rid of `get_inference_dynamic_batching_max_tokens()` in the example script, since we're not ready for this method quite yet. Let me know if it looks okay

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 2 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference.py]: I like it!

File: megatron/core/inference/contexts/dynamic_context.py (hunk 3 above)
Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py]: I went through, this looks good!

**Comment by kvareddy** [comment]
@sidsingh-nvidia @santhnm2 @shanmugamr1992 @kanz-nv can you please take a look at this MR? It simplifies dynamic context in a significant way.

**Review by santhnm2 — APPROVED** [review]
LGTM but this looks like it will conflict with the Mamba changes [!1907](https://github.com/NVIDIA/Megatron-LM/pull/1907) (nothing fundamental but slightly non-trivial merge will be required)

File: megatron/core/inference/contexts/dynamic_context.py [hunk 5]
```diff
@@ -207,26 +217,19 @@ class DynamicInferenceContext(BaseInferenceContext):
         num_attention_heads (int): Number of attention heads.
         max_sequence_length (int): Max possible sequence length (prompt + output)
             that will occur.
-        buffer_size_gb (float): Total buffer size (GB), shared by main and
-            fallback contexts.
+        active_buffer_size_gb (float): Buffer size reserved for active requests
+            that live on the GPU. The total buffer size (stored in unified memory)
+            is 2x this value, with the the other half of the buffer reserved for
+            paused requests that live on the CPU.
+        max_tokens (int): Max number of tokens to use for forward passes. This is
```
Comment (kvareddy) [inline megatron/core/inference/contexts/dynamic_context.py:229]: How does the max_tokens argument work with chunked prefill support?

Comment (lmcafee-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py:229]: I haven't had any issues running with chunked prefill. Is there something I should check?
